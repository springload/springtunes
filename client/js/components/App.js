import React, { Component, PropTypes } from 'react';
import LastUpdated from '../components/LastUpdated';
import IntelligentCurrentSong from '../containers/IntelligentCurrentSong';
import IntelligentControls from '../containers/IntelligentControls';
import Logo from '../components/Logo';

const propTypes = {
    lastUpdated: PropTypes.number.isRequired,
    fetchSongIfNeeded: PropTypes.func.isRequired,
};

class App extends Component {
    componentDidMount() {
        this.props.fetchSongIfNeeded();
    }

    render() {
        const { lastUpdated } = this.props;
        return (
            <div>
                <div className="header">
                    <div className="header__logo container">
                        <Logo />
                    </div>
                </div>
                <div className="container section">
                    <div className="last-updated">
                        <p className="u-text-uppercase">Now Playing</p>
                        <LastUpdated lastUpdated={lastUpdated} />
                    </div>
                    <IntelligentCurrentSong />
                    <IntelligentControls />
                </div>
            </div>
        );
    }
}

App.propTypes = propTypes;

export default App;
