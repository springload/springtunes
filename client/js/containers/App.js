import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { fetchSongIfNeeded, togglePauseIfNeeded, nextIfNeeded, backIfNeeded,
muteIfNeeded, unmuteIfNeeded, changeVolume } from '../actions';
import LastUpdated from '../components/LastUpdated';
import Controls from '../components/Controls';
import CurrentSong from '../components/CurrentSong';
import Logo from '../components/Logo';

class App extends Component {
    constructor(props) {
        super(props);
        this.handleRefreshClick = this.handleRefreshClick.bind(this);
        this.handleTogglePauseClick = this.handleTogglePauseClick.bind(this);
        this.handleNextClick = this.handleNextClick.bind(this);
        this.handleBackClick = this.handleBackClick.bind(this);
        this.handleMuteClick = this.handleMuteClick.bind(this);
        this.handleUnmuteClick = this.handleUnmuteClick.bind(this);
        this.handleVolumeChange = this.handleVolumeChange.bind(this);
    }

    componentDidMount() {
        const { dispatch } = this.props;
        dispatch(fetchSongIfNeeded());
    }

    handleRefreshClick(e) {
        e.preventDefault();

        const { dispatch } = this.props;
        dispatch(fetchSongIfNeeded());
    }

    handleTogglePauseClick(e) {
        e.preventDefault();

        const { dispatch } = this.props;
        dispatch(togglePauseIfNeeded());
    }

    handleNextClick(e) {
        e.preventDefault();

        const { dispatch } = this.props;
        dispatch(nextIfNeeded());
    }

    handleBackClick(e) {
        e.preventDefault();

        const { dispatch } = this.props;
        dispatch(backIfNeeded());
    }

    handleMuteClick(e) {
        e.preventDefault();

        const { dispatch } = this.props;
        dispatch(muteIfNeeded());
    }

    handleUnmuteClick(e) {
        e.preventDefault();

        const { dispatch } = this.props;
        dispatch(unmuteIfNeeded());
    }

    handleVolumeChange(volume) {
        const { dispatch } = this.props;
        dispatch(changeVolume(volume));
    }

    render() {
        const { currentSong, isFetching, playing, volume, lastUpdated, isModifyingMute, isMuted, error } = this.props;
        return (
            <div>
				<div className='header'>
                    <div className='header__logo container'>
                        <Logo />
                    </div>
                </div>
                <div className='container section'>
                    <div className='last-updated'>
                    <p className='u-text-uppercase'>Now Playing</p>
                    <LastUpdated lastUpdated={ lastUpdated } />
                </div>
                <CurrentSong isFetching={ isFetching } currentSong={ currentSong } error={ error } />
                <Controls isFetching={ isFetching }
                  playing={ playing }
                  handleRefreshClick={ this.handleRefreshClick }
                  handleTogglePauseClick={ this.handleTogglePauseClick }
                  handleNextClick={ this.handleNextClick }
                  handleBackClick={ this.handleBackClick }
                  isMuted={ isMuted }
                  isModifyingMute={ isModifyingMute }
                  handleMuteClick={ this.handleMuteClick }
                  handleUnmuteClick={ this.handleUnmuteClick }
                  volume={ volume }
                  handleVolumeChange={ this.handleVolumeChange }
                />
            </div>
		</div>
        );
    }
}

App.propTypes = {
    currentSong: PropTypes.object.isRequired,
    isFetching: PropTypes.bool.isRequired,
    playing: PropTypes.bool.isRequired,
    volume: PropTypes.number,
    error: PropTypes.string,
    lastUpdated: PropTypes.number,
    isModifyingMute: PropTypes.bool.isRequired,
    isMuted: PropTypes.bool.isRequired,
    dispatch: PropTypes.func.isRequired,
};

function mapStateToProps(state) {
    const { songManager, volumeManager, errorManager } = state;
    let isFetching = songManager.current ? songManager.current.isFetching : true;
    const playing = songManager.current ? songManager.current.playing : false;
    let volume = songManager.current && songManager.current.volume ? songManager.current.volume : 50;
    let lastUpdated = songManager.current ? songManager.current.lastUpdated : Date.now();
    const currentSong = songManager.current ? songManager.current.currentSong : {};

    const isModifyingMute = volumeManager.muting ? volumeManager.muting.isModifyingMute : false;
    const isMuted = volumeManager.muting ? volumeManager.muting.isMuted : false;

    const error = errorManager.error ? errorManager.error.error : '';
    if (error !== '') {
        isFetching = false;
        lastUpdated = errorManager.error ? errorManager.error.lastUpdated : lastUpdated;
    }

    if (volumeManager.volume) {
        volume = volumeManager.volume ? volumeManager.volume.value : volume;
    }

    return {
        currentSong,
        isFetching,
        playing,
        volume,
        lastUpdated,
        isModifyingMute,
        isMuted,
        error,
    };
}

export default connect(mapStateToProps)(App);
