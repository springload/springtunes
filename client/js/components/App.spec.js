import React from 'react';
import renderer from 'react-test-renderer';
import App from './App';

jest.mock('./Logo', () => 'div');
jest.mock('./LastUpdated', () => 'div');
jest.mock('../containers/IntelligentCurrentSong', () => 'div');
jest.mock('../containers/IntelligentControls', () => 'div');

describe('App component', () => {
    it('renders correctly with min params', () => {
        const result = renderer.create(
            <App lastUpdated={23525345345} fetchSongIfNeeded={() => {}} />,
        );
        expect(result).toMatchSnapshot();
    });

    it('fetchSong is called', () => {
        let fetchSong = false;
        const shallowRenderer = renderer.create(
            <App
                lastUpdated={23525345345}
                fetchSongIfNeeded={() => {
                    fetchSong = true;
                }}
            />,
        );
        const instance = shallowRenderer.getInstance();
        instance.componentDidMount();
        expect(fetchSong).toBeTruthy();
    });
});
