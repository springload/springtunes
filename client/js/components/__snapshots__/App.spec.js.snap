exports[`App component renders correctly with min params 1`] = `
<div>
  <div
    className="header">
    <div
      className="header__logo container">
      <Unknown />
    </div>
  </div>
  <div
    className="container section">
    <div
      className="last-updated">
      <p
        className="u-text-uppercase">
        Now Playing
      </p>
      <Unknown
        lastUpdated={23525345345} />
    </div>
    <Connect(Component) />
    <Connect(Component) />
  </div>
</div>
`;
